Define actor data

Ampersandrew opened the pull request:
Define the data models for the actors and items.

Ampersandrew committed:
Start defining the data model

Move the data model stuff out of sheets into data. Sheets is for sheets,
not data models.

Define the data model for characters and NPCs

Ampersandrew committed:
Start defining the data model

Move the data model stuff out of sheets into data. Sheets is for sheets,
not data models.

Define the data model for characters and NPCs

Ampersandrew committed:
Tidy up unneeded file

Ampersandrew committed:
Update the character data definitions

A bit of refactoring.

Ampersandrew committed:
Also changed somestuff in here during the refactor

Ampersandrew committed:
Clone the socketlib stuff into the system

To avoid having external dependencies the socketlib library
has been cloned into our system. The MIT lisence allows for
this.

Ampersandrew committed:
Bugfix Sleeving not undoable

Sleeving a card was not storing enough state to allow it to be
undone.

While fixing that noticed that nextTurn, Vamoose and sleeveCard
had a lot of common code, refactored the common stuff. As part of
this changed the sleeveHightest socket operation to not take
a combatantId as only the acting combatant can sleeve a card.

Also noticed that the now extracted common saveTurn function
had a lot in common with previousTurn. Extracted the common
bits into another function.

Removed some redundant stuff in the socket implementation.

Ampersandrew committed:
Oops fix merge conflict

Ampersandrew committed:
Add in chip logic

Ampersandrew committed:
Tidy up

Move stuff to better locations.

Change the language file for better consistency.

Ampersandrew committed:
Merge remote-tracking branch 'upstream/develop' into DefineActorData

bring the actor data branch up to date with develop

Ampersandrew committed:
Update data models

clean up the dlc fields.

Split items to make edges and hinderances separate from physical items.
clean up the datamodels surrounding this.

Ampersandrew committed:
Merge branch 'Dulux-Oz:develop' into DefineActorData

TheSpyder commented on module/data/dlc-fields.mjs:
This is definitely where TypeScript would help, to make sure all these object shapes line up correctly. But I don't think we want to tackle that just yet.

TheSpyder commented on module/data/dlc-fields.mjs:
```diff
@@ -0,0 +1,217 @@
+  initialMax = 10,
+  required = true,
+  isInteger = true,
+  ...options
```
`options` isn't used for many of the definitions in this file?

Ampersandrew replied:
True, I was iterating and I think that's vestigial.

TheSpyder commented on module/data/chip-data.mjs:
```diff
@@ -0,0 +1,19 @@
+
+    return {
+      chips: new fields.SchemaField({
+        ...dlcFields.integerNoMax('white', 0, 0),
+        ...dlcFields.integerNoMax('red', 0, 0),
+        ...dlcFields.integerNoMax('blue', 0, 0),
```
Today I learned... the marshall doesn't get green chips

TheSpyder commented on module/deadlands-classic.mjs:
```diff
@@ -56,11 +71,26 @@ Hooks.once('init', async () => {
-  await fpPreloadTemplates();
+  if (!Array.isArray(globalThis.game.chips)) {
+    game.chips = {};
+    game.chips.apps = {};
```
Should it be array or object? `Array.isArray` on those values will return false.

Ampersandrew replied:
It should be an object. I have corrected this.

TheSpyder commented on module/helpers/chips.mjs:
```diff
@@ -0,0 +1,68 @@
+    } else if (picked <= white + red + blue) {
+      game.chips.available.blue -= 1;
+      pick = Chips.type.Blue;
+    } else {
```
```suggestion
    } else if (includeGreen) {
```

Just in case there's a logic flaw, we wouldn't want green to be returned by accident 😂

Ampersandrew replied:
I committed this.

TheSpyder commented on module/init/add-chip-tab.mjs:
```diff
@@ -0,0 +1,50 @@
+  icon.setAttribute('class', `fas fa-yin-yang`);
+  tab.append(icon);
+
+  // Add Chip Manager tab to sidebar before itemss if it's not already there
```
```suggestion
  // Add Chip Manager tab to sidebar before items if it's not already there
```

TheSpyder commented on module/init/add-chip-tab.mjs:
```diff
@@ -0,0 +1,50 @@
+export function addChipTab(app, html) {
+  if (!game.user.isGM) return;
+
+  // Calculate new tab width
+  html[0]
+    .querySelector('#sidebar-tabs')
+    .style.setProperty(
+      '--sidebar-tab-width',
+      `${Math.floor(
+        parseInt(
+          getComputedStyle(html[0]).getPropertyValue('--sidebar-width'),
+          10
+        ) /
+          (document.querySelector('#sidebar-tabs').childElementCount + 1)
+      )}px`
+    );
```
That was way too much in a template value, and normally I'd extract a few more variables out of the `floor` expression too. But one step at a time.

```suggestion
  const computedWidth = Math.floor(
    parseInt(getComputedStyle(html[0]).getPropertyValue('--sidebar-width'), 10)
     /
    (document.querySelector('#sidebar-tabs').childElementCount + 1)
  )
  // Calculate new tab width
  html[0]
    .querySelector('#sidebar-tabs')
    .style.setProperty('--sidebar-tab-width', `${computedWidth}px`);
```

Ampersandrew replied:
I grabbed this off the internet.

TheSpyder commented on templates/icons/svg/card-joker.svg:
This file is zero bytes, what happened?

TheSpyder reviewed (approved):
Sorry this took me a month to get to!

I don't have the data model in my head, but it all looks pretty good from a quick scan.

Ampersandrew committed:
Correct check of Array.isArray

Ampersandrew committed:
A bit of validation

Add a check for if (includeGreen) to ensure the marshal isn't given a
green chip.

Ampersandrew committed:
Update module/init/add-chip-tab.mjs

Co-authored-by: Andrew Herron <thespyder@programmer.net>

Ampersandrew committed:
Update module/init/add-chip-tab.mjs

Co-authored-by: Andrew Herron <thespyder@programmer.net>

TheSpyder commented:
github tip: if you browse through the suggestions using the "files changed" tab, you can collect multiple suggestions into a single commit :)

Ampersandrew commented:
Thanks

Can you tell I don't use it a lot?

-- 
  Andrew Wilson
  ***@***.***


On Wed, 1 May 2024, at 2:09 AM, Andrew Herron wrote:
> 
> 
> github tip: if you browse through the suggestions using the "files changed" tab, you can collect multiple suggestions into a single commit :)
> 
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/Dulux-Oz/DeadlandsClassic/pull/11#issuecomment-2087798905>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABDARTQ7LOSGVM7TMHQ2SKTZAA6ERAVCNFSM6AAAAABFDJVGQCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAOBXG44TQOJQGU>.
> You are receiving this because you modified the open/close state.Message ID: ***@***.***>
>
